include/graphblas/reference/blas3.hpp

```diff
@@ -1204,6 +1204,210 @@ namespace grb {
 			return SUCCESS;
 		}
 
+		template<
+			bool upper,
+			Descriptor descr = descriptors::no_operation,
+			typename InputType, typename OutputType,
+			typename RIT_L, typename CIT_L, typename NIT_L,
+			typename RIT_A, typename CIT_A, typename NIT_A
+		>
+		RC trilu_generic(
+			Matrix< OutputType, reference, RIT_L, CIT_L, NIT_L > &L,
+			const Matrix< InputType, reference, RIT_A, CIT_A, NIT_A > &A,
+			const long int k,
+			const Phase &phase
+		) {
+#ifdef _DEBUG
+			std::cout << "In grb::internal::trilu_generic( reference )\n";
+#endif
+			const size_t m = descr & descriptors::transpose_matrix
+							? ncols( A ) : nrows( A );
+			const size_t n = descr & descriptors::transpose_matrix
+							? nrows( A ) : ncols( A );
+			const size_t nz = nnz( A );
+
+			if( m == 0 || n == 0 || nz == 0 ) {
+#ifdef _DEBUG
+				std::cout << "Empty matrix, nothing to compute.\n";
+#endif
+				return SUCCESS;
+			}
+
+			// Run-time checks
+			if( m != nrows( L ) || n != ncols( L ) ) {
+#ifdef _DEBUG
+				std::cout << "Mismatched dimensions detected.\n";
+#endif
+				return MISMATCH;
+			}
+
+			// Check that there is no overlap between L and A
+			if( getID(A) == getID(L) ) {
+#ifdef _DEBUG
+				std::cout << "Overlap detected between input and output matrices.\n";
+#endif
+				return OVERLAP;
+			}
+
+			const auto &A_raw = descr & descriptors::transpose_matrix
+								? internal::getCCS( A )
+								: internal::getCRS( A );
+
+			if( phase == Phase::RESIZE ) {
+				size_t nzc = 0;
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+#pragma omp parallel for reduction( + : nzc ) \
+		default( none ) shared( A_raw ) firstprivate( k, m )
+#endif
+				for( size_t i = 0; i < m; ++i ) {
+					const auto A_k_start = A_raw.col_start[ i ];
+					const auto A_k_end = A_raw.col_start[ i + 1 ];
+					for( auto A_k = A_k_start; A_k < A_k_end; ++A_k ) {
+						const auto A_j = A_raw.row_index[ A_k ];
+
+						// If the value is in the specified triangle, skip it
+						const long A_j_long = static_cast< long >( A_j );
+						const long i_long = static_cast< long >( i );
+						if( ( !upper && A_j_long > i_long + k  )
+							|| ( upper && A_j_long + k < i_long )
+						) {
+							continue;
+						}
+
+						nzc += 1;
+					}
+				}
+#ifdef _DEBUG
+				std::cout << "RESIZE phase: resize( L, " << nzc << " )\n";
+#endif
+				return resize( L, nzc );
+			}
+
+			if( phase == Phase::EXECUTE ) {
+				auto &L_crs_raw = internal::getCRS( L );
+				auto &L_ccs_raw = internal::getCCS( L );
+				const size_t nzc = capacity( L );
+
+				L_crs_raw.col_start[ 0 ] = 0;
+				L_ccs_raw.col_start[ 0 ] = 0;
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+#pragma omp parallel for simd
+#endif
+				for( size_t i = 0; i < m; ++i ) {
+					L_ccs_raw.col_start[ i + 1 ] = 0;
+				}
+
+				// Prefix sum computation into L.CRS.col_start
```

Comment:
Is this similar to the various other prefix sums elsewhere in the code base (e.g., `reference/matrix.hpp`? At this point it is probably wise to have like a single standard function for it in `graphblas/utils` e.g.

Reply:
Already working on that 👍